oeps/best-practices/oep-0051-bp-conventional-commits.rst

```diff
@@ -88,6 +88,8 @@ We use these commit types labels:
 
 * **revert**: undo a previous change. The previous commit message, including the type label, is included.
 
+* **squash**: temporary changes that are intended to be squashed before merging.
```

Comment:
This seems un-necessary.  Git already has a notion of indicating that a commit is meant to be squashed into another(`squash! <title of commit to squash into>`), and that commit title already fails our lint testing so it achieves both the communicating of intent(better than the squash type in my opinion since the title can be used to map back to where it was meant to be squashed) and fails tests like the new squash type would. 

The value of making this change seems to be that "temp feels wrong" when we want to indicate other information?  Perhaps what we need is to say that normal git notation around fixups and squashes are allowed in PRs but we should aim to resolve them before putting commits onto our main branches.

Reply:
1. Can you point to docs for "squash!"? I'm not familiar with it.
2. Your comment "This seems un-necessary" seems at odds with your help with how you would do this. Maybe you are saying it is unnecessary to invent a new method of doing this, and we should instead document that we want people to follow this git convention?
3. My original intent was to have a convention that both conveys meaning **and** got me a green build. I'll admit that without the promise of a green build, this doesn't seem worth pursuing to me. And as noted in my pros/cons, I think there are trade-offs with pass/fail on this, and don't think failing is the only valid solution.
4. I'll admit that I was happily using "temp" and "fix" for these commits, and I didn't think I was doing anything wrong, and I was happy to have a green build as I wanted. I now see that if you read the definition of "temp", then you would know that was not its intent. For "fix", these temporary commits often are fixing something on the PR. Even "refactor" would work for some of these commits. Note that if I fix or refactor something in a follow-up PR, these are exactly the labels I would use, and the question is whether those are only valid labels in a follow-up PR, and not in an in-progress PR.

Reply:
1. https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---squashltcommitgt
2. Yes, I think it's un-necessary to invent a new way.
3/4. I think the goal of these labels is to help make the commits clear.  It's up to the rest of the words in the commit message to support this initial word and you can always be more clear in those words.  I think it's fine to use fix/refactor in the same PR as the thing that they are fixing.  Commits don't know about PRs and are just referencing things that happened earlier in the commit history.

Reply:
Certainly a fair take.  On the other hand, I like:
- Keeping and working with the CC format habits
- Using a title directly related to this (sub-) change, rather than the original commit's title
- I think this will make it easier to write a custom commitlint rule that can warn for these without blocking if we decide to make the commitlint check "required" but still allow squash-and-merge, which could make that transition more palatable.

Reply:
Thanks @feanil and @inventhouse. I agree with @inventhouse's response. 

For those developers that use `autosquash`, I'm not sure if the tooling (commit linting, auto change logs, etc.) could also handle that well, and maybe that could be taken up as an additional alternative path for those that want to use `autosquash` **and** want a green build. If you are ok with a failed build, you can really use whatever you want.

Reply:
I was also realizing this morning that git already has customs and tooling for detailed squashing (N->M squashes).  @robrap/@inventhouse I don't know what style of commit messages you are envisioning here.  How much detail is in the "squash:" messages?

Reply:
@nedbat honestly probably not a lot, usually just `squash:` and a title.  Also, I'm likely to experiment with the git tooling that was referenced above.

My main interest in this is around flagging commits that are not intended to be merged as-written and especially helping folks use the tooling to track that; I'd be quite happy with just a mention that using git's tooling and formatting or one's own convention on development branches is a good way to do this with the linting we do today and discourage folks from using `temp`, etc. for changes that are not meant to be temporary or commits that are not meant to be merged.

If/when we discuss ratcheting-up the commitlint check to be required, we can discuss / investigate custom rules for non-blocking warnings and such if we like.

Reply:
@nedbat: At this time, as someone who already uses conventional commits, I find the failures annoying and will likely not look at them because they are just crying wolf about the squash commits I already know about. If I happen to mis-type or misremember one, and there is a real error, I probably will miss it.

I am open to any formatting for squash commits that would quiet the linter. If we aren’t quieting the linter, none of this really matters much to me. 🙂

Reply:
With what I said above, I would be interested in pursuing custom warning rules to help developers track the commits they do not intend to merge as-written _while also_ catching genuine errors and reducing the incentive to mis-use accepted type tags.

Reply:
https://github.com/edx/edx-lint/pull/215 would make "squash" a warning, and not an error, which you can see in action on that PR. 

Are there any disagreements with this direction?

FYI: @nedbat @inventhouse

Reply:
It looks like @timmc-edx also demonstrated that commitlint is already configured to ignore `squash!` (i.e. it passes **and** provides no warnings). 

I propose we just document behavior of `squash!` and allow it and move on. This simplifies docs and custom code.

I would drop the separate `squash:` type and custom `edx-lint` code because we would just have more stuff to maintain.

Thoughts?